fix: Provide a way to customize the graceful shutdown delay of Netty `EventLoopGroup`s

novoj opened the pull request:
Added parameters to enable setting the quiet period and timeout for graceful shutdown of worker groups, both globally and through the `ClientFactoryBuilder`.

Make use of graceful shutdown settings on worker groups on the server side.

Refs: #5813

Motivation:

My E2E integration tests are running slowly just because they wait for default graceful shutdown timeout to expire.

Modifications:

- made use of existing graceful shutdown settings on the server side
- added new graceful (missing) shutdown settings on the client side

Result:

- Closes #5813.
- this pull request has no consequences for the user (#IMHO) - as a default values I configured the defaults that are applied by Netty default implementation
- user can now explicitly configure zero timeout and quiet period both for server and client side, which results in immediate closing of both and speeding up the termination phase considerably

novoj commented on core/src/main/java/com/linecorp/armeria/client/ClientFactoryBuilder.java:
```diff
@@ -926,6 +926,50 @@ public ClientFactoryBuilder connectionOutlierDetection(OutlierDetection outlierD
+        option(ClientFactoryOptions.WORKER_GROUP_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS,
+               workerGroupGracefulShutdownTimeoutMillis);
+        return this;
+    }
```
Ok, I'd add this annotation.

novoj commented on core/src/main/java/com/linecorp/armeria/common/Flags.java:
```diff
@@ -1076,6 +1084,40 @@ public static long defaultServerConnectionDrainDurationMicros() {
+     */
+    public static long defaultClientWorkerGroupGracefulShutdownTimeoutMillis() {
+        return DEFAULT_CLIENT_WORKER_GROUP_GRACEFUL_SHUTDOWN_TIMEOUT_MILLIS;
+    }
```
Ok, I'll add it.

trustin commented:
I like the approach here and the feedback from @jrhee17 and @ikhoon. Thanks for bringing back to the table, @novoj! 🙇

jrhee17 commented:
I think it's best to start with the direct issue you would like to solve - specifying `GracefulShutdown` for `ClientFactory#workerGroup`.

What do you think of starting with adding the following APIs in ClientFactoryBuilder.java, ClientFactoryOption.java?

```java ClientFactoryBuilder.java
    public ClientFactoryBuilder workerGroup(EventLoopGroup workerGroup, boolean shutdownOnClose,
                                                                          GracefulShutdown gracefulShutdown) {
    }

    public ClientFactoryBuilder workerGroup(int numThreads, GracefulShutdown gracefulShutdown) {
    }
```

```java ClientFactoryOption.java
    public static final ClientFactoryOption<GracefulShutdown> WORKER_GROUP_GRACEFUL_SHUTDOWN =
            ClientFactoryOption.define("WORKER_GROUP_GRACEFUL_SHUTDOWN", Flags.workerGroupGracefulShutdown());
```

novoj commented:
Here is what I did:

1. I removed the logic inside the Server class because that's where most of the reservations and suggestions for changes were.
2. I updated (synced) my branch with the current main branch.
3. I polished some JavaDoc.

The client-side implementation remained the same. I think the actual handling is much more aligned with existing methods, such as:

- `com.linecorp.armeria.common.FlagsProvider#defaultClientWorkerGroupGracefulShutdownTimeoutMillis`
- `com.linecorp.armeria.common.FlagsProvider#defaultClientHttp2GracefulShutdownTimeoutMillis`

The newly proposed methods by @jrhee17 in the comment at https://github.com/line/armeria/pull/6528#issuecomment-3604790436 suit the server-side API better, in my opinion. The current pull request is minimalistic and still solves my problem. It seems that the delays in my tests are caused by the client rather than the server shutdown.

This patch is consistent and can be merged separately from the server-side solution, which I may attempt in a separate pull request.
